README.md
# I2C Character LCD Driver

![Screen](images/IMG_2554.jpg)

Provides an embedded Rust driver for common 16x2 LCD displays that use the AiP31068L chip to
drive the display, and a PCA9633 chip to drive the RGB backlight.

This has been tested with the [Waveshare LCD1602 module](https://www.waveshare.com/wiki/LCD1602_RGB_Module).
It may also work with other RGB displays like the [Groove 16X2 LDC RGB](https://www.seeedstudio.com/Grove-LCD-RGB-Backlight-p-1643.html),
but I haven't tested it.

This is a basic implementation, and doesn't currently support custom characters.

## Speed

This driver is fast enough that there are no noticable delays when updating text on the screen. In
my programs, I'm using spaces to clear the ends of lines rather than clearing the screen.

## Example

Currently there is a single, but working, example of using this crate in the [examples/STM32F10x](examples/STM32F10x) folder.
The example is using an STM32F103RB Nucleo-64 board. This is a nice little board as it has the programmer/debugger
built into the board.

```rust
let scl = gpiob.pb10.into_alternate_open_drain(&mut gpiob.crh);
let sda = gpiob.pb11.into_alternate_open_drain(&mut gpiob.crh);

let i2c_bus = BlockingI2c::i2c2(
    peripherals.I2C2,
    (scl, sda),
    i2c::Mode::Standard {
        frequency: 400_000.hz(),
    },
    clocks,
    &mut rcc.apb1,
    1000,
    10,
    1000,
    1000,
);

let mut lcd = Lcd::new(i2c_bus, LCD_ADDRESS, RGB_ADDRESS, &mut delay).unwrap();
lcd.set_rgb(255, 255, 255).unwrap();
lcd.print("Hello world!").unwrap();
```

## License

This project is licensed under MIT license ([LICENSE](https://github.com/kunerd/clerk/blob/master/docs/CONTRIBUTING.md) or <https://opensource.org/licenses/MIT>)

CHANGELOG.md
# Change Logs

## [0.3.0] 2021-08-07

This release represents a breaking change designed to make the API more idiomatic. For example, instead
of using `set_cursor_on()` and `set_cursor_off()` methods, it now uses `set_cursor(Cursor::On)` and
`set_cursor(Cursor:Off)`.

tests/integration.rs
use embedded_hal_mock::common::Generic;
use embedded_hal_mock::delay::MockNoop;
use embedded_hal_mock::i2c::{Mock as I2cMock, Transaction as I2cTransaction};
use lcd_1602_i2c::{Lcd, Cursor, LcdDisplay, Blink};
use std::vec;

const BLINK_ON: u8 = 0x01;
const CURSOR_ON: u8 = 0x02;
const DISPLAY_ON: u8 = 0x04;
const DISPLAY_CONTROL: u8 = 0x08;
const EXPECTED_ADDRESS: u8 = 123;
const RGB_ADDRESS: u8 = 34;

// Ensure the initialization sequence doesn't break
#[test]
fn lcd_initialization() {
    let expectations = lcd_expectations(EXPECTED_ADDRESS, RGB_ADDRESS);
    let _ = new_lcd(&expectations, EXPECTED_ADDRESS, RGB_ADDRESS);
}

#[test]
fn display_off() {
    // Arrange
    let expectations =
        lcd_expectations(EXPECTED_ADDRESS, RGB_ADDRESS)
        .command_bytes(EXPECTED_ADDRESS, DISPLAY_CONTROL);

    let mut lcd = new_lcd(&expectations, EXPECTED_ADDRESS, RGB_ADDRESS);

    // Act
    let _ = lcd.set_display(LcdDisplay::Off);
}

#[test]
fn cursor_on() {
    // Arrange
    let expectations =
        lcd_expectations(EXPECTED_ADDRESS, RGB_ADDRESS)
        .command_bytes(EXPECTED_ADDRESS, DISPLAY_CONTROL | DISPLAY_ON | CURSOR_ON);

    let mut lcd = new_lcd(&expectations, EXPECTED_ADDRESS, RGB_ADDRESS);

    // Act
    let _ = lcd.set_cursor(Cursor::On);
}

#[test]
fn cursor_off() {
    // Arrange
    let expectations =
        lcd_expectations(EXPECTED_ADDRESS, RGB_ADDRESS)
        .command_bytes(EXPECTED_ADDRESS, DISPLAY_CONTROL | DISPLAY_ON | CURSOR_ON)
        .command_bytes(EXPECTED_ADDRESS, DISPLAY_CONTROL | DISPLAY_ON);

    let mut lcd = new_lcd(&expectations, EXPECTED_ADDRESS, RGB_ADDRESS);
    let _ = lcd.set_cursor(Cursor::On);

    // Act
    let _ = lcd.set_cursor(Cursor::Off);
}

#[test]
fn blink_on() {
    // Arrange
    let expectations =
        lcd_expectations(EXPECTED_ADDRESS, RGB_ADDRESS)
        .command_bytes(EXPECTED_ADDRESS, DISPLAY_CONTROL | DISPLAY_ON | BLINK_ON);

    let mut lcd = new_lcd(&expectations, EXPECTED_ADDRESS, RGB_ADDRESS);

    // Act
    let _ = lcd.set_blink(Blink::On);
}

#[test]
fn blink_off() {
    // Arrange
    let expectations =
        lcd_expectations(EXPECTED_ADDRESS, RGB_ADDRESS)
        .command_bytes(EXPECTED_ADDRESS, DISPLAY_CONTROL | DISPLAY_ON | BLINK_ON)
        .command_bytes(EXPECTED_ADDRESS, DISPLAY_CONTROL | DISPLAY_ON);

    let mut lcd = new_lcd(&expectations, EXPECTED_ADDRESS, RGB_ADDRESS);
    let _ = lcd.set_blink(Blink::On);

    // Act
    let _ = lcd.set_blink(Blink::Off);
}

fn new_lcd<'a>(
    expectations: &'a Expectations,
    address: u8,
    rgb_address: u8,
) -> Lcd<Generic<I2cTransaction>> {
    let mut delay = MockNoop::new();
    let i2c = I2cMock::new(expectations.as_array());
    Lcd::new(i2c, address, rgb_address, &mut delay).unwrap()
}

// Returns expectations for the initialization of the LCD display that is always
// via new().
fn lcd_expectations(address: u8, rgb_address: u8) -> Expectations {
    Expectations::new()
        // Initial command sequence for "wakeup"
        .command_bytes(address, 0x28)
        .command_bytes(address, 0x28)
        .command_bytes(address, 0x28)
        // Display On
        .command_bytes(address, 8 | 4)
        // Clear
        .command_bytes(address, 0x01)
        // Set LCD mode
        .command_bytes(address, 0x02 | 0x04)
        // Initialize the RGB backlight
        .reg_bytes(rgb_address, 0, 0)
        .reg_bytes(rgb_address, 8, 0xFF)
        .reg_bytes(rgb_address, 1, 0x20)
}

struct Expectations {
    expectations: Vec<I2cTransaction>,
}

impl Expectations {
    pub fn new() -> Self {
        Expectations {
            expectations: Vec::new(),
        }
    }

    pub fn command_bytes(mut self, address: u8, byte: u8) -> Self {
        self.expectations
            .push(I2cTransaction::write(address, vec![0x80, byte]));
        self
    }

    fn reg_bytes(mut self, address: u8, reg: u8, byte: u8) -> Self {
        self.expectations
            .push(I2cTransaction::write(address, vec![reg, byte]));
        self
    }

    pub fn as_array(&self) -> &[I2cTransaction] {
        self.expectations.as_slice()
    }
}


src/display_control.rs
#![deny(missing_docs)]

/// Controls the visibilty of the non-blinking cursor, which is basically an _ **after** the cursor position.
/// The cursor position represents where the next character will show up.
#[derive(Copy, Clone)]
pub enum Cursor {
    /// Display the non-blinking cursor
    On = 2,
    /// Hide the non-blinking cursor
    Off = 0,
}

/// Determines whether the entire LCD is on or off.
#[derive(Copy, Clone)]
pub enum LcdDisplay {
    /// Turn the LCD display on
    On = 4,
    /// Turn the LCD display off
    Off = 0,
}

/// Controls the visibility of the blinking block cursor.
#[derive(Copy, Clone)]
pub enum Blink {
    /// Turn the blinking block cursor on
    On = 1,
    /// Turn the blinking block cursor off
    Off = 0,
}

pub struct DisplayControl {
    pub cursor: Cursor,
    pub display: LcdDisplay,
    pub blink: Blink,
}

impl DisplayControl {
    pub fn new() -> Self {
        DisplayControl {
            cursor: Cursor::Off,
            display: LcdDisplay::Off,
            blink: Blink::Off,
        }
    }

    pub fn value(&self) -> u8 {
        0x08 | self.blink as u8 | self.cursor as u8 | self.display as u8
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn all_flags_off() {
        let control = DisplayControl::new();
        assert_eq!(0x08, control.value());
    }

    #[test]
    fn cursur_on() {
        let mut control = DisplayControl::new();
        control.cursor = Cursor::On;
        assert_eq!(0x08 | 2, control.value());
    }

    #[test]
    fn blink_on() {
        let mut control = DisplayControl::new();
        control.blink = Blink::On;
        assert_eq!(0x08 | 1, control.value());
    }

    #[test]
    fn display_on() {
        let mut control = DisplayControl::new();
        control.display = LcdDisplay::On;
        assert_eq!(0x08 | 4, control.value());
    }

    #[test]
    fn display_and_cursor_on() {
        let mut control = DisplayControl::new();
        control.display = LcdDisplay::On;
        control.cursor = Cursor::On;
        assert_eq!(0x08 | 4 | 2, control.value());
    }
}


examples/STM32F10x/src/main.rs
// src/main.rs

#![no_std]
#![no_main]

use cortex_m_rt::entry; // The runtime

use hal::{
    delay::Delay,
    i2c::{self, BlockingI2c},
    pac,
    prelude::*,
};
use stm32f1xx_hal as hal; // STM32F1 specific functions

use lcd_1602_i2c::{Lcd, Cursor};

#[allow(unused_imports)]
use panic_halt; // When a panic occurs, stop the microcontroller

#[entry]
fn main() -> ! {
    let peripherals = pac::Peripherals::take().unwrap();
    let core_peripherals = cortex_m::Peripherals::take().unwrap();

    let mut rcc = peripherals.RCC.constrain();

    let mut gpiob = peripherals.GPIOB.split(&mut rcc.apb2);

    let mut flash = peripherals.FLASH.constrain();

    let clocks = rcc.cfgr.sysclk(8.mhz()).freeze(&mut flash.acr);

    let mut delay = Delay::new(core_peripherals.SYST, clocks);

    // Now setup the LCD display
    let scl = gpiob.pb10.into_alternate_open_drain(&mut gpiob.crh);
    let sda = gpiob.pb11.into_alternate_open_drain(&mut gpiob.crh);

    let i2c_bus = BlockingI2c::i2c2(
        peripherals.I2C2,
        (scl, sda),
        i2c::Mode::Standard {
            frequency: 400_000.hz(),
        },
        clocks,
        &mut rcc.apb1,
        1000,
        10,
        1000,
        1000,
    );

    let mut lcd = Lcd::new(i2c_bus, LCD_ADDRESS, RGB_ADDRESS, &mut delay).unwrap();
    lcd.set_rgb(255, 255, 255).unwrap();
    lcd.set_cursor(Cursor::On).unwrap();
    lcd.write_str("Hello world!").unwrap();

    loop {
    }
}

// Device I2c addresses
const LCD_ADDRESS: u8 = 0x7c >> 1;
const RGB_ADDRESS: u8 = 0xc0 >> 1;


src/lib.rs
#![deny(missing_docs)]

/*!
# Platform-agnostic driver for I2C 16x2 character displays

Provides a driver for common 16x2 LCD displays that use the AiP31068L chip to
drive the display, and a PCA9633 chip to drive the RGB backlight.

This is a basic implementation, and doesn't currently support custom characters.

This has been tested with the [Waveshare LCD1602 module](https://www.waveshare.com/wiki/LCD1602_RGB_Module).
It may also work with other RGB displays like the [Groove 16X2 LDC RGB](https://www.seeedstudio.com/Grove-LCD-RGB-Backlight-p-1643.html)
*/

#![no_std]
use embedded_hal::blocking::{i2c, delay::DelayMs};

mod display_control;
use display_control::{DisplayControl};

pub use display_control::{Cursor, LcdDisplay, Blink};

/**
Handles all the logic related to working with the character LCD via I2C. You'll
need to create an instance of this with the `new()` method.

The `I` generic type needs to implement the `embedded_hal::blocking::Write` trait.
*/
pub struct Lcd<I>
where
    I: i2c::Write,
{
    i2c: I,
    control: DisplayControl,
    address: u8,
    rgb_address: u8,
}

impl<I> Lcd<I>
where
    I: i2c::Write
    {
    /**
    Creates a new instance of the display object.

    # Example

    ```rust
    let lcd = Lcd::new(i2c_bus, address, rgb_address, &mut delay);
    ```

    `i2c` needs to implement the `embedded_hal::blocking::Write` trait.

    `delay` needs to implement the `embedded_hal::blocking::delay::DelayMs` trait.

    # Errors

    The I2C library will return an error if it's not able to write to the device.
    This is always a trait of type `embedded_hal::blocking::Write::Error` that
    is implemented by the I2C instance.
    */
    pub fn new<D>(i2c: I, address: u8, rgb_address: u8, delay: &mut D) -> Result<Self, <I as i2c::Write>::Error>
    where
        D: DelayMs<u16>
    {
        let mut display = Lcd {
            i2c,
            control: DisplayControl::new(),
            address,
            rgb_address,
        };
        display.init(delay)?;
        Ok(display)
    }

    // Initialize the display for the first time after power up
    fn init<D>(&mut self, delay: &mut D) -> Result<(), <I as i2c::Write>::Error>
    where D: DelayMs<u16> {
        delay.delay_ms(80); // Need to wait at least 40ms before sending commands

        // Send the initial command sequence according to the HD44780 datasheet
        self.write_function_set()?;
        delay.delay_ms(5);

        self.write_function_set()?;
        delay.delay_ms(5);

        self.write_function_set()?;

        self.set_display(LcdDisplay::On)?;

        self.clear(delay)?;

        // Display entry mode
        const LCD_ENTRYLEFT: u8 = 0x02;
        const LCD_ENTRYSHIFTDECREMENT: u8 = 0x00;
        const LCD_ENTRYMODESET: u8 = 0x04;

        self.write_command(LCD_ENTRYMODESET | LCD_ENTRYLEFT | LCD_ENTRYSHIFTDECREMENT)?;

        // Initialize the backlight
        const REG_MODE1: u8     = 0x00;
        const REG_MODE2: u8     = 0x01;
        const REG_OUTPUT: u8    = 0x08;
    
        self.write_reg(REG_MODE1, 0)?;

        // Set the LEDs controllable by both PWM and GRPPWM registers
        self.write_reg(REG_OUTPUT, 0xFF)?;
        self.write_reg(REG_MODE2, 0x20)
    }

    /**
    Clear the display. The LCD display driver requires a 2ms delay after clearing, which
    is why this method requires a `delay` object.

    # Errors

    Returns a `Result` that will report I2C errors, if any.
    */
    pub fn clear(&mut self, delay: &mut dyn DelayMs<u16>) -> Result<(), <I as i2c::Write>::Error> {
        const LCD_CLEARDISPLAY: u8 = 0x01;

        let result = self.write_command(LCD_CLEARDISPLAY);
        delay.delay_ms(2);
        result
    }

    /**
    Set the position of the cursor

    # Errors

    Returns a `Result` that will report I2C errors, if any.
    */
    pub fn set_cursor_position(&mut self, x: u8, y: u8) -> Result<(), <I as i2c::Write>::Error> {
        let col = if y == 0_u8 { x | 0x80 } else { x | 0xC0 };
        self.write_command(col)
    }

    /**
    Control whether the display is on or off

    # Errors

    Returns a `Result` that will report I2C errors, if any.
    */
    pub fn set_display(&mut self, display: LcdDisplay) -> Result<(), <I as i2c::Write>::Error> {
        self.control.display = display;
        self.write_display_control()
    }

    /**
    Sets the visiblity of the cursor, which is a non-blinking _

    # Errors

    Returns a `Result` that will report I2C errors, if any.
    */
    pub fn set_cursor(&mut self, cursor: Cursor) -> Result<(), <I as i2c::Write>::Error> {
        self.control.cursor = cursor;
        self.write_display_control()
    }

    /**
    Turns on the blinking block cursor

    # Errors

    Returns a `Result` that will report I2C errors, if any.
    */
    pub fn set_blink(&mut self, blink: Blink) -> Result<(), <I as i2c::Write>::Error> {
        self.control.blink = blink;
        self.write_display_control()
    }

    /**
    Adds a single character to the current position. The cursor will advance
    after this call to the next column

    # Errors

    Returns a `Result` that will report I2C errors, if any.
    */
    pub fn write_char(&mut self, char: char) -> Result<(), <I as i2c::Write>::Error> {
        self.write_two(0x40, char as u8)
    }

    /**
    Adds a string to the current position. The cursor will advance
    after this call to the next column

    # Errors

    Returns a `Result` that will report I2C errors, if any.
    */
    pub fn write_str(&mut self, s: &str) -> Result<(), <I as i2c::Write>::Error> {
        for c in s.chars() {
            self.write_char(c)?;
        }

        Ok(())
    }

    /**
    Set the color of the backlight for displays that have an RGB backlight.

    # Errors

    Returns a `Result` that will report I2C errors, if any.
    */
    pub fn set_rgb(&mut self, r: u8, g: u8, b: u8) -> Result<(), <I as i2c::Write>::Error> {
        const REG_RED: u8       = 0x04;        // pwm2
        const REG_GREEN: u8     = 0x03;        // pwm1
        const REG_BLUE: u8      = 0x02;        // pwm0
    
        self.write_reg(REG_RED, r)?;
        self.write_reg(REG_GREEN, g)?;
        self.write_reg(REG_BLUE, b)
    }

    fn write_reg(&mut self, addr: u8, data: u8) -> Result<(), <I as i2c::Write>::Error> {
        self.i2c.write(self.rgb_address, &[addr, data])
    }

    fn write_function_set(&mut self) -> Result<(), <I as i2c::Write>::Error> {
        const LCD_4BITMODE: u8 = 0x00;
        const LCD_2LINE: u8 = 0x08;
        const LCD_5X8_DOTS: u8 = 0x00;
        const LCD_FUNCTIONSET: u8 = 0x20;

        self.write_command(LCD_FUNCTIONSET | LCD_4BITMODE | LCD_2LINE | LCD_5X8_DOTS)
    }

    // Set one of the display's control options and then send the updated set of options to the display
    fn write_display_control(&mut self) -> Result<(), <I as i2c::Write>::Error> {
        self.write_command(self.control.value())
    }

    // Send a command to the LCD display
    fn write_command(&mut self, value: u8) -> Result<(), <I as i2c::Write>::Error> {
        self.write_two(0x80, value)
    }

    // Send two bytes to the display
    fn write_two(&mut self, byte1: u8, byte2: u8) -> Result<(), <I as i2c::Write>::Error> {
        self.i2c.write(self.address, &[byte1, byte2])
    }
}
